misc/amia_paper_2024/stats.py

```diff
@@ -0,0 +1,417 @@
+
+
+def setup():
+    """Setup any environmental prerequisites"""
+    # handles like:
+    # SettingWithCopyWarning:
+    # A value is trying to be set on a copy of a slice from a DataFrame.
+    # Try using .loc[row_indexer,col_indexer] = value instead
+    # See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
+    #   df_get_csets['codeset_ids'] = df_get_csets['codeset_ids'].astype(str)
+    pd.options.mode.chained_assignment = None  # default='warn'
+
+    if not INDIR.exists():
+        OUTDIR.mkdir()
+    if not OUTDIR.exists():
+        OUTDIR.mkdir()
+
+
+def api_runs_query(verbose=False):
+    """Get all records from api_runs table"""
+    t0 = datetime.now()
+    with get_db_connection() as con:
+        data: List[RowMapping] = sql_query(
+            con, """
+                SELECT DISTINCT *,
+                date_bin('1 week', timestamp::TIMESTAMP, TIMESTAMP '2023-10-30')::date week,
+                timestamp::date date FROM public.api_runs r""")
+    data: List[Dict] = [dict(d) for d in data]
+    if verbose:
+        print(f'api_runs_query(): Fetched {len(data)} records in n seconds: {(datetime.now() - t0).seconds}')
+    return data
+
+
+def histogram_query_concepts_in_calls(filter_nulls=True) -> pd.DataFrame:
+    """Concepts in calls
+
+    result, modified (NULLs removed, at the very least)
+    ┌─────────────────────┬───────┐
+    │  concepts_in_call   │ calls │
+    ├─────────────────────┼───────┤
+    │ 1 - 9               │    20 │
+    │ 10 - 99             │  1,642 │
+    │ 100 - 999           │   855 │
+    │ 1000 - 9999         │   505 │
+    │ 10000 - 99999       │   153 │
+    │ 100000 - 999999     │   121 │
+    │ 1000000 - 9999999   │     3 │
+    │ 10000000 - 99999999 │     2 │
+    └─────────────────────┴───────┘
+    """
+    # Query
+    qry = """
+        WITH c1 AS (
+          SELECT replace(result, ' rows', '') cnt FROM public.apijoin
+          WHERE api_call IN ('codeset-ids-by-concept-id', 'concept', 'get-cset-members-items')
+        ), c2 AS (
+          SELECT CASE WHEN cnt ~ '^[0-9]+$' THEN cnt::integer END AS cnt FROM c1
+        )
+        SELECT 10^round(log10(cnt)) || ' - ' || 10 * 10^round(log10(cnt)) - 1 concepts_in_call, count(*) calls
+        FROM c2 GROUP BY 1 ORDER BY 1 DESC;"""
+    with get_db_connection() as con:
+        data: List[RowMapping] = sql_query(con, qry)
+    data: List[Dict] = [dict(d) for d in data]
+    df = pd.DataFrame(data)
+    # Filter
+    if filter_nulls:
+        df = df[df['concepts_in_call'].notna()]
+    # Format
+    def relabel(val):
+        """Relabel column vals: log10"""
+        min_val = val.split(' - ')[0]
+        n = min_val.count('0')
+        return f'10^{n}-10^{n+1}'
+    df['concepts_in_call'] = df['concepts_in_call'].apply(relabel)
+    return df
+
+
+# todo: y axis also in log scale
+def plot_concepts_in_calls():
```

Comment:
## Plot: Concepts in calls
Here's what I got.

For the small version, I wasn't sure the best way to do the xlabels and still get them to fit in a small space. I've included 2 variations.

I also realize that the y data is too small in some bins to be shown. This can be fixed with some more logarithmic scaling, but I'll move onto other plots and I can come back to this if we have time.

![n_concepts_in_api_calls - small](https://github.com/jhu-bids/TermHub/assets/13045020/f3496027-f728-4650-848c-154d8279976a)

![n_concepts_in_api_calls - small2](https://github.com/jhu-bids/TermHub/assets/13045020/7b7bfcd9-c373-457a-8358-f0c4304e8930)

![n_concepts_in_api_calls](https://github.com/jhu-bids/TermHub/assets/13045020/b9f6c4b3-0937-4c73-bc3c-ce0c745fdc78)